pallets/phala/src/fat.rs

```diff
@@ -0,0 +1,284 @@
+	#[pallet::storage]
+	pub type ContractCode<T: Config> = StorageMap<_, Twox64Concat, CodeHash<T>, Vec<u8>>;
+
+	/// The contract cluster counter.
+	#[pallet::storage]
+	pub type ContractClusterCounter<T> = StorageValue<_, u64, ValueQuery>;
+
+	#[pallet::storage]
+	pub type ContractClusters<T> = StorageMap<_, Twox64Concat, ContractClusterId, Vec<ContractId>>;
+
+	#[pallet::storage]
+	pub type Contracts<T: Config> =
+		StorageMap<_, Twox64Concat, ContractId, ContractInfo<CodeHash<T>, T::AccountId>>;
+
+	#[pallet::storage]
+	pub type ClusterWorkers<T> =
+		StorageMap<_, Twox64Concat, ContractClusterId, Vec<WorkerPublicKey>>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		CodeUploaded(CodeHash<T>),
+		PubkeyAvailable(ContractId, ContractPublicKey),
+		Instantiating(ContractId, ContractClusterId, T::AccountId),
+		Instantiated(ContractId, ContractClusterId, H256),
+		InstantiationFailed(ContractId, ContractClusterId, H256),
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		CodeNotFound,
+		ContractClusterNotFound,
+		DuplicatedContract,
+		DuplicatedDeployment,
+		NoWorkerSpecified,
+		InvalidSender,
+		WorkerNotFound,
+	}
+
+	type CodeHash<T> = <T as frame_system::Config>::Hash;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T>
+	where
+		T: crate::mq::Config + crate::registry::Config,
+	{
+		#[pallet::weight(0)]
+		pub fn upload_code(origin: OriginFor<T>, code: Vec<u8>) -> DispatchResult {
+			ensure_signed(origin)?;
+			let code_hash = T::Hashing::hash(&code);
+			ContractCode::<T>::insert(&code_hash, &code);
+			Self::deposit_event(Event::CodeUploaded(code_hash));
+			Ok(())
+		}
+
+		#[pallet::weight(0)]
+		pub fn upload_code_to_cluster(
+			origin: OriginFor<T>,
+			code: Vec<u8>,
+			cluster_id: ContractClusterId,
+		) -> DispatchResult {
+			let origin: T::AccountId = ensure_signed(origin)?;
+			// TODO.shelven: check permission?
+			Self::push_message(ContractOperation::UploadCodeToCluster {
+				origin,
+				code,
+				cluster_id,
+			});
+			Ok(())
+		}
+
+		#[pallet::weight(0)]
+		pub fn instantiate_contract(
+			origin: OriginFor<T>,
+			// #[pallet::compact] endowment: BalanceOf<T>,
+			// #[pallet::compact] gas_limit: Weight,
+			code_index: CodeIndex<CodeHash<T>>,
+			data: Vec<u8>,
+			salt: Vec<u8>,
+			deploy_to: DeployTarget,
+		) -> DispatchResult {
+			let deployer = ensure_signed(origin)?;
+
+			match code_index {
+				CodeIndex::NativeCode(_) => {}
+				CodeIndex::WasmCode(code_hash) => {
+					ensure!(
+						ContractCode::<T>::contains_key(code_hash),
+						Error::<T>::CodeNotFound
+					);
+				}
+			}
+
+			let (cluster_id, deploy_workers) = match deploy_to {
+				DeployTarget::Cluster(cluster_id) => {
+					ensure!(
+						ClusterWorkers::<T>::contains_key(cluster_id),
+						Error::<T>::ContractClusterNotFound
+					);
+					let workers = ClusterWorkers::<T>::get(cluster_id).expect("checked; qed.");
```

Comment:
2 read ops and the later one is uncencessay. The db ops are very expensive in Substrate. So instead we should write like this:

```rust
let workers = ClusterWorkers::<T>::get(cluster_id).ok_or(Error::<T>::ContractClusterNotFound)?;
```

Reply:
Fixed.